llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -1585,6 +1585,26 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(ISD::MSTORE, VT, Custom);
     }
 
+    if (EnablePartialReduceNodes) {
+      for (MVT VT : MVT::integer_scalable_vector_valuetypes()) {
+        for (MVT InnerVT : MVT::integer_scalable_vector_valuetypes()) {
+          // 1. Set all combinations where a type is illegal to "Legal"
+          // - These will be legalized to a legal type pair
+          // - Avoid expanding them too early (or preventing folds)
+          if (!isTypeLegal(VT) || !isTypeLegal(InnerVT)) {
+            setPartialReduceMLAAction(VT, InnerVT, Legal);
+            continue;
+          }
+          //  2. Set all legal combinations to "Expand"
+          // - Not all of these can be lowered (via a Legal or Custom lowering).
+          setPartialReduceMLAAction(VT, InnerVT, Expand);
+        }
+      }
+      // 3. Mark known legal pairs as 'Legal' (these will expand to USDOT).
```

Comment:
In the table, should we also encode the kind of extension being done? Because for SVE the USDOT instruction only supports i8 -> i32, whereas for SDOT it supports i16 -> i32, i8 -> i32 and i16 -> i64.

Reply:
The comment here is referring to SDOT or UDOT not USDOT (my bad). I don't think we should include extension info, as the node does not include that information (so it would not be available where `getPartialReduceMLAAction` is used without looking at operands UUIC). I think the USDOT case can be handled by marking applicable type pairs as "Custom", the custom lowering can then handle producing USDOTs when applicable. I believe if a custom lowering just returns "SDValue()" the legalizer assumes the node is legal, so we can still use the standard ISEL for non-USDOTs.

Reply:
Yes, we'll need to do custom lowering for USDOT because it's probably a bit too target-specific to warrant creating a generic ISD node for it. That said, I think there is value in encoding the extensions as well so that it can be used by the cost-model/TTI (see `TargetTransformInfo::getPartialReductionCost`, which takes a `PartialReductionExtendKind` for both operands) to figure out whether or not such an operation is legal/custom for the given types and given extensions. But that doesn't necessarily need doing in this PR.